### ISSUE TITLE
MEMORY INEFFICIENCY: Parallel cluster workers hold on to objects longer than necessary

### ISSUE THREAD

**Issue by HenrikBengtsson** [HenrikBengtsson/Wishlist-for-R#27]
# Issue
The internal `parallel:::slaveLoop()` function that acts as the main program for cluster workers created by `parallel::makeCluster()` hangs on to the object returned to the master process longer than necessary.

## Example
With
```r
## Memory usage of a process (only Linux)
pmem <- function(pid) {
  res <- system2("pmap", args=pid, stdout=TRUE)
  res <- grep("total", res, value=TRUE)
  res <- strsplit(res, split="[ ]+", res)[[1]]
  grep("[0-9]+", res, value=TRUE)
}
```
we get the following
```r
> library("parallel")
> cl <- makeCluster(1L)

## Get the PID for the worker
> pid <- clusterEvalQ(cl, { Sys.getpid() })[[1]]
> pid
[1] 16087
> pmem(pid)
[1] "203696K"

## Have worker allocate large object and return
> res <- clusterEvalQ(cl, { integer(length = 100e6) })
> object.size(res)
400000088 bytes
> pmem(pid)
[1] "594324K"

## Ask worker to garbage collect (above memory is NOT released)
> res <- clusterEvalQ(cl, { gc() })
> pmem(pid)
[1] "594324K"

## Ask worker to garbage collect again (above memory IS released)
> res <- clusterEvalQ(cl, { gc() })
> pmem(pid)
[1] "203696K"
````

Note how the first garbage collection has little effect and it's only in the second call that the `value` object from the first iteration is removed.

# Troubleshooting
The reason for the _first_ garbage collection to have no effect is that `parallel:::slaveLoop()` holds on to the value of the first call in local variable `value` throughout the evaluation of the second call as well, which only overwrites the local `value` variable after completing the evaluation.  This can be seen if one study [its `repeat` statement](https://github.com/wch/r-source/blob/tags/R-3-3-1/src/library/parallel/R/worker.R#L21-L50) below:
```r
slaveLoop <- function(master)
{
    repeat
        tryCatch({
            msg <- recvData(master)
            # cat(paste("Type:", msg$type, "\n"))

            if (msg$type == "DONE") {
                closeNode(master)
                break;
            } else if (msg$type == "EXEC") {
                success <- TRUE
                ## This uses the message rather than the exception since
                ## the exception class/methods may not be available on the
                ## master.
                handler <- function(e) {
                    success <<- FALSE
                    structure(conditionMessage(e),
                              class = c("snow-try-error","try-error"))
                }
                t1 <- proc.time()
                value <- tryCatch(do.call(msg$data$fun, msg$data$args, quote = TRUE),
                                  error = handler)
                t2 <- proc.time()
                value <- list(type = "VALUE", value = value, success = success,
                              time = t2 - t1, tag = msg$data$tag)
                sendData(master, value)
            }
        }, interrupt = function(e) NULL)
}
```

# Patch
The obvious solution is be to remove local variable `value` _directly after_ the data has been sent back to the master process.  Analogously, the local variable `msg` should also be removed as soon as possible, because it may contain large data objects.  That one can be removed _before_ the data is sent back to the master process, which then would be available for garbage collection while `sendData()` is called.

## Subversion patch
```sh
$ svn diff src/library/parallel/R/worker.R
```
```r
Index: src/library/parallel/R/worker.R
===================================================================
--- src/library/parallel/R/worker.R	(revision 70874)
+++ src/library/parallel/R/worker.R	(working copy)
@@ -44,7 +44,9 @@
                 t2 <- proc.time()
                 value <- list(type = "VALUE", value = value, success = success,
                               time = t2 - t1, tag = msg$data$tag)
+                rm(list = "msg")
                 sendData(master, value)
+                rm(list = "value")
             }
         }, interrupt = function(e) NULL)
 }
```

This patch has been validated by rebuilding R from SVN source and running `make check`.

With this patch, we get:
```r
> library("parallel")
> cl <- makeCluster(1L)
> pid <- clusterEvalQ(cl, { Sys.getpid() })[[1]]
> pmem(pid)
[1] "235832K"

> res <- clusterEvalQ(cl, { integer(length = 100e6) })
> pmem(pid)
[1] "626460K"

> res <- clusterEvalQ(cl, { gc() })
> pmem(pid)
[1] "235832K"
```

**Comment by HenrikBengtsson** [comment]
I've submitted the above patch, cf. [PR #17115](https://bugs.r-project.org/bugzilla3/show_bug.cgi?id=17115) ('MEMORY: More memory efficient SNOW worker [PATCH]').

**Comment by HenrikBengtsson** [comment]
Patch applied:

[Luke Tierney 2016-10-13 18:39:08 UTC](https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17115#c2):
> Committed to [R-devel in r71517](https://github.com/wch/r-source/commit/bfd3f38a29d0879b266295cc4966fff3d121725a) and [R-patched in 71518](https://github.com/wch/r-source/commit/e4bbebb82d9e0472b0f92751e8ed90890086b3de), using more compiler-friendly NULL assignment instead of rm().

In other words, this will be available in upcoming R 3.3.2.

Case closed.
